File: BankingWebsite.Tests/Steps/LoginPageSteps.cs
using Microsoft.Playwright;
using BankingWebsite.Tests.PageObjects;
using BankingWebsite.Tests.Constants;
using TechTalk.SpecFlow;
using FluentAssertions;

namespace BankingWebsite.Tests.Steps
{
    [Binding]
    public class LoginPageSteps
    {
        private readonly IPage _page;
        private readonly LoginPage _loginPage;

        public LoginPageSteps(IPage page)
        {
            _page = page;
            _loginPage = new LoginPage(_page);
        }

        [Given(@"I am on the login choices page")]
        public async Task GivenIAmOnTheLoginChoicesPage()
        {
            await _loginPage.NavigateToLoginPage();
            var isLoaded = await _loginPage.IsLoginPageLoaded();
            isLoaded.Should().BeTrue("Login page should be loaded successfully");
        }

        [When(@"I click on the ""([^""]*)"" link")]
        [When(@"I click on the ""([^""]*)"" login option")]
        public async Task WhenIClickOnTheLink(string linkText)
        {
            switch (linkText.ToLower())
            {
                case "login":
                    var homePage = new HomePage(_page);
                    await homePage.ClickLoginLink();
                    break;
                case "personal":
                    await _loginPage.ClickPersonalLoginLink();
                    break;
                case "business":
                    await _loginPage.ClickBusinessLoginLink();
                    break;
                case "client fund service":
                    await _loginPage.ClickClientFundServiceLink();
                    break;
                case "wealthnet":
                    await _loginPage.ClickWealthNetLink();
                    break;
                case "support":
                    await _loginPage.ClickSupportLink();
                    break;
                case "contact":
                    await _loginPage.ClickContactLink();
                    break;
                default:
                    throw new ArgumentException($"Link '{linkText}' is not implemented in step definitions");
            }
        }

        [When(@"I navigate to the login choices page")]
        public async Task WhenINavigateToTheLoginChoicesPage()
        {
            await _loginPage.NavigateToLoginPage();
        }

        [Then(@"I should be redirected to the login choices page")]
        public async Task ThenIShouldBeRedirectedToTheLoginChoicesPage()
        {
            var currentUrl = await _loginPage.GetCurrentUrl();
            currentUrl.Should().Be(TestConstants.LOGIN_CHOICES_URL, "Should be on the login choices page");
        }

        [Then(@"I should see the ""([^""]*)"" heading")]
        public async Task ThenIShouldSeeTheHeading(string headingText)
        {
            switch (headingText.ToLower())
            {
                case "internet banking login":
                    var isHeadingVisible = await _loginPage.IsInternetBankingLoginHeadingVisible();
                    isHeadingVisible.Should().BeTrue($"'{headingText}' heading should be visible");
                    break;
                case "choose type of internet banking":
                    var isChooseTypeVisible = await _loginPage.IsChooseTypeHeadingVisible();
                    isChooseTypeVisible.Should().BeTrue($"'{headingText}' heading should be visible");
                    break;
                default:
                    throw new ArgumentException($"Heading '{headingText}' verification is not implemented");
            }
        }

        [Then(@"I should see the ""([^""]*)"" text")]
        public async Task ThenIShouldSeeTheText(string expectedText)
        {
            switch (expectedText.ToLower())
            {
                case "choose type of internet banking":
                    var isVisible = await _loginPage.IsChooseTypeHeadingVisible();
                    isVisible.Should().BeTrue($"Text '{expectedText}' should be visible");
                    break;
                default:
                    // For other text verification, we can add more specific locators
                    throw new ArgumentException($"Text '{expectedText}' verification is not implemented");
            }
        }

        [Then(@"I should see ""([^""]*)"" and ""([^""]*)"" login options")]
        public async Task ThenIShouldSeeAndLoginOptions(string option1, string option2)
        {
            var isPersonalVisible = await _loginPage.IsPersonalLoginLinkVisible();
            var isBusinessVisible = await _loginPage.IsBusinessLoginLinkVisible();
            
            isPersonalVisible.Should().BeTrue($"'{option1}' login option should be visible");
            isBusinessVisible.Should().BeTrue($"'{option2}' login option should be visible");
        }

        [Then(@"I should see the login page title ""([^""]*)""")]
        public async Task ThenIShouldSeeTheLoginPageTitle(string expectedTitle)
        {
            var actualTitle = await _loginPage.GetPageTitle();
            actualTitle.Should().Be(expectedTitle, $"Login page title should be '{expectedTitle}'");
        }

        [Then(@"I should see the ""([^""]*)"" section")]
        public async Task ThenIShouldSeeTheSection(string sectionName)
        {
            switch (sectionName.ToLower())
            {
                case "personal or business":
                    var isPersonalVisible = await _loginPage.IsPersonalLoginLinkVisible();
                    var isBusinessVisible = await _loginPage.IsBusinessLoginLinkVisible();
                    isPersonalVisible.Should().BeTrue("Personal login link should be visible");
                    isBusinessVisible.Should().BeTrue("Business login link should be visible");
                    break;
                default:
                    throw new ArgumentException($"Section '{sectionName}' verification is not implemented");
            }
        }

        [Then(@"I should see the ""([^""]*)"" login link")]
        public async Task ThenIShouldSeeTheLoginLink(string linkType)
        {
            switch (linkType.ToLower())
            {
                case "personal":
                    var isPersonalVisible = await _loginPage.IsPersonalLoginLinkVisible();
                    isPersonalVisible.Should().BeTrue("Personal login link should be visible");
                    break;
                case "business":
                    var isBusinessVisible = await _loginPage.IsBusinessLoginLinkVisible();
                    isBusinessVisible.Should().BeTrue("Business login link should be visible");
                    break;
                default:
                    throw new ArgumentException($"Login link type '{linkType}' verification is not implemented");
            }
        }

        [Then(@"I should be redirected to the personal internet banking login page")]
        public async Task ThenIShouldBeRedirectedToThePersonalInternetBankingLoginPage()
        {
            await Task.Delay(2000); // Wait for navigation
            var currentUrl = await _loginPage.GetCurrentUrl();
            currentUrl.Should().Contain("secure.bnz.co.nz/foyer/target/ib", "Should be on personal internet banking login page");
        }

        [Then(@"I should be redirected to the business internet banking login page")]
        public async Task ThenIShouldBeRedirectedToTheBusinessInternetBankingLoginPage()
        {
            await Task.Delay(2000); // Wait for navigation
            var currentUrl = await _loginPage.GetCurrentUrl();
            currentUrl.Should().Contain("secure.bnz.co.nz/foyer/target/ib4b", "Should be on business internet banking login page");
        }

        [Then(@"the URL should contain ""([^""]*)""")]
        [Then(@"I should be redirected to a page with URL containing ""([^""]*)""")]
        public async Task ThenTheURLShouldContain(string expectedUrlPart)
        {
            await Task.Delay(2000); // Wait for navigation
            var currentUrl = await _loginPage.GetCurrentUrl();
            currentUrl.Should().Contain(expectedUrlPart, $"URL should contain '{expectedUrlPart}'");
        }
    }
}

File: Tests/Steps/LoginPageSteps.ts
import { Given, When, Then } from '@cucumber/cucumber';
import { expect } from '@playwright/test';
import { LoginPage } from '../PageObjects/LoginPage';
import { TestConstants } from '../Constants/TestConstants';
import { page } from '../Utilities/BrowserContext';

let loginPage: LoginPage;

// Given steps - Preconditions
Given('I am on the login choices page', async function () {
    loginPage = new LoginPage(page);
    await page.goto(TestConstants.LOGIN_CHOICES_URL);
    await loginPage.waitForLoginChoicesPageLoad();
});

// When steps - Actions
When('I click on the {string} login option', async function (loginType: string) {
    loginPage = loginPage || new LoginPage(page);
    await loginPage.clickLoginOption(loginType);
});

When('I click on the Personal login option', async function () {
    loginPage = loginPage || new LoginPage(page);
    await loginPage.clickPersonalLogin();
});

When('I click on the Business login option', async function () {
    loginPage = loginPage || new LoginPage(page);
    await loginPage.clickBusinessLogin();
});

When('I click on the Client Fund Service option', async function () {
    loginPage = loginPage || new LoginPage(page);
    await loginPage.clickClientFundService();
});

When('I click on the WealthNet option', async function () {
    loginPage = loginPage || new LoginPage(page);
    await loginPage.clickWealthNet();
});

When('I select the {string} login type from the options', async function (loginType: string) {
    loginPage = loginPage || new LoginPage(page);
    
    switch (loginType.toLowerCase()) {
        case 'personal':
            await loginPage.clickPersonalLogin();
            break;
        case 'business':
            await loginPage.clickBusinessLogin();
            break;
        default:
            throw new Error(`Unsupported login type: ${loginType}`);
    }
});

// Then steps - Assertions
Then('I should be redirected to the login choices page', async function () {
    await page.waitForLoadState('networkidle');
    loginPage = loginPage || new LoginPage(page);
    
    const currentURL = await loginPage.getCurrentURL();
    expect(currentURL).toContain('login-choices');
});

Then('I should see the page title {string}', async function (expectedTitle: string) {
    await page.waitForLoadState('domcontentloaded');
    loginPage = loginPage || new LoginPage(page);
    
    const actualTitle = await loginPage.getCurrentTitle();
    expect(actualTitle).toBe(expectedTitle);
});

Then('I should see the heading {string}', async function (expectedHeading: string) {
    loginPage = loginPage || new LoginPage(page);
    
    const headingText = await loginPage.getHeadingText();
    expect(headingText).toContain(expectedHeading);
});

Then('I should see login options for {string} and {string}', async function (option1: string, option2: string) {
    loginPage = loginPage || new LoginPage(page);
    
    const loginOptionsVisible = await loginPage.verifyLoginOptions(option1, option2);
    expect(loginOptionsVisible).toBeTruthy();
    
    // Additional specific checks
    const personalVisible = await loginPage.verifyPersonalLoginOption();
    const businessVisible = await loginPage.verifyBusinessLoginOption();
    
    expect(personalVisible).toBeTruthy();
    expect(businessVisible).toBeTruthy();
});

Then('I should be redirected to the {string} login page', async function (loginType: string) {
    await page.waitForLoadState('networkidle');
    
    const currentURL = page.url();
    
    switch (loginType.toLowerCase()) {
        case 'personal':
            expect(currentURL).toContain('secure.bnz.co.nz/foyer/target/ib');
            break;
        case 'business':
            expect(currentURL).toContain('secure.bnz.co.nz/foyer/target/ib4b');
            break;
        default:
            throw new Error(`Unknown login type: ${loginType}`);
    }
});

Then('the URL should contain {string}', async function (expectedUrlPart: string) {
    const currentURL = page.url();
    expect(currentURL).toContain(expectedUrlPart);
});

Then('I should see alternative login options', async function () {
    loginPage = loginPage || new LoginPage(page);
    
    const alternativeOptionsVisible = await loginPage.verifyAlternativeLoginOptions();
    expect(alternativeOptionsVisible).toBeTruthy();
});

Then('I should see {string} option', async function (optionName: string) {
    loginPage = loginPage || new LoginPage(page);
    
    const optionVisible = await loginPage.verifySpecificAlternativeOption(optionName);
    expect(optionVisible).toBeTruthy();
});

Then('the login page should have proper accessibility elements', async function () {
    loginPage = loginPage || new LoginPage(page);
    
    const hasAccessibilityElements = await loginPage.verifyAccessibilityElements();
    expect(hasAccessibilityElements).toBeTruthy();
});

Then('the page should have a main heading', async function () {
    loginPage = loginPage || new LoginPage(page);
    
    const hasMainHeading = await loginPage.verifyLoginPageHeading();
    expect(hasMainHeading).toBeTruthy();
});

Then('navigation elements should be accessible', async function () {
    loginPage = loginPage || new LoginPage(page);
    
    const hasNavigation = await loginPage.verifyElementVisible('nav[role="navigation"], header nav');
    expect(hasNavigation).toBeTruthy();
});

// Parameterized and reusable step definitions
Then('I should see the login page element {string}', async function (elementSelector: string) {
    loginPage = loginPage || new LoginPage(page);
    
    const elementVisible = await loginPage.verifyElementVisible(elementSelector);
    expect(elementVisible).toBeTruthy();
});

Then('I should see text {string} on the login page', async function (expectedText: string) {
    const textLocator = page.locator(`text=${expectedText}`);
    await expect(textLocator).toBeVisible();
});

Then('I should not see text {string} on the login page', async function (unexpectedText: string) {
    const textLocator = page.locator(`text=${unexpectedText}`);
    await expect(textLocator).not.toBeVisible();
});

Then('I should see the login page URL is {string}', async function (expectedURL: string) {
    const currentURL = page.url();
    expect(currentURL).toBe(expectedURL);
});

Then('I should see the login page URL contains {string}', async function (urlPart: string) {
    const currentURL = page.url();
    expect(currentURL).toContain(urlPart);
});

Then('I wait for the login page to load completely', async function () {
    loginPage = loginPage || new LoginPage(page);
    await loginPage.waitForLoginChoicesPageLoad();
});

// Verification steps for specific login types
Then('I should see Personal login option is clickable', async function () {
    loginPage = loginPage || new LoginPage(page);
    
    const personalLogin = loginPage.getPersonalLoginLink();
    await expect(personalLogin).toBeVisible();
    await expect(personalLogin).toBeEnabled();
});

Then('I should see Business login option is clickable', async function () {
    loginPage = loginPage || new LoginPage(page);
    
    const businessLogin = loginPage.getBusinessLoginLink();
    await expect(businessLogin).toBeVisible();
    await expect(businessLogin).toBeEnabled();
});

// Steps for verifying multiple elements at once
Then('I should see all primary login options', async function () {
    loginPage = loginPage || new LoginPage(page);
    
    const personalVisible = await loginPage.verifyPersonalLoginOption();
    const businessVisible = await loginPage.verifyBusinessLoginOption();
    
    expect(personalVisible).toBeTruthy();
    expect(businessVisible).toBeTruthy();
});

Then('I should see all alternative login services', async function () {
    loginPage = loginPage || new LoginPage(page);
    
    const clientFundVisible = await loginPage.verifySpecificAlternativeOption('Client Fund Service');
    const wealthNetVisible = await loginPage.verifySpecificAlternativeOption('WealthNet');
    
    expect(clientFundVisible).toBeTruthy();
    expect(wealthNetVisible).toBeTruthy();
});